source/lib.ts

```diff
@@ -0,0 +1,264 @@
+					: options.max
+
+			const maxHits = await Promise.resolve(retrieveQuota)
+			// Set the rate limit information on the augmented request object
+			augmentedRequest[options.requestPropertyName] = {
+				limit: maxHits,
+				current: totalHits,
+				remaining: Math.max(maxHits - totalHits, 0),
+				resetTime,
+			}
+
+			// Set the X-RateLimit headers on the response object if enabled
+			if (options.headers && !response.headersSent) {
+				response.setHeader('X-RateLimit-Limit', maxHits)
+				response.setHeader(
+					'X-RateLimit-Remaining',
+					augmentedRequest[options.requestPropertyName].remaining,
+				)
+
+				// If we have a resetTime, also provide the current date to help avoid issues with incorrect clocks
+				if (resetTime instanceof Date) {
+					response.setHeader('Date', new Date().toUTCString())
+					response.setHeader(
+						'X-RateLimit-Reset',
+						Math.ceil(resetTime.getTime() / 1000),
+					)
+				}
+			}
+
+			// Set the standardized RateLimit headers on the response object
+			// if enabled
+			if (options.useStandardizedHeaders && !response.headersSent) {
+				response.setHeader('RateLimit-Limit', maxHits)
+				response.setHeader(
+					'RateLimit-Remaining',
+					augmentedRequest[options.requestPropertyName].remaining,
+				)
+
+				if (resetTime) {
+					const deltaSeconds = Math.ceil(
+						(resetTime.getTime() - Date.now()) / 1000,
+					)
+					response.setHeader('RateLimit-Reset', Math.max(0, deltaSeconds))
+				}
+			}
+
+			// If we are to skip failed/successfull requests, decrement the
+			// counter accordingly once we know the status code of the request
+			if (options.skipFailedRequests || options.skipSuccessfulRequests) {
+				let decremented = false
+				const decrementKey = () => {
+					if (!decremented) {
+						options.store.decrement(key)
+						decremented = true
+					}
+				}
+
+				if (options.skipFailedRequests) {
+					response.on('finish', () => {
+						if (!options.requestWasSuccessful(request, response)) decrementKey()
+					})
+					response.on('close', () => {
+						if (!response.writableEnded) decrementKey()
+					})
+					response.on('error', () => {
+						decrementKey()
+					})
+				}
+
+				if (options.skipSuccessfulRequests) {
+					response.on('finish', () => {
+						if (options.requestWasSuccessful(request, response)) decrementKey()
+					})
+				}
+			}
+
+			// Call the {@link Options.onLimitReached} callback if
+			// the client has reached their rate limit.
+			if (maxHits && totalHits === maxHits + 1) {
+				options.onLimitReached(request, response, options)
+			}
+
+			// If the client has exceeded their rate limit, set the Retry-After
+			// header and call the {@link Options.handler} function
+			if (maxHits && totalHits > maxHits) {
+				if (options.headers && !response.headersSent) {
+					response.setHeader('Retry-After', Math.ceil(options.windowMs / 1000))
+				}
+
+				options.handler(request, response, options)
+				return
+			}
+
+			next()
+		},
+	)
+
```

Comment:
Probably add that init call right here. Something like this:

```suggestion
if (typeof options.store.init === 'function') {
  options.store.init(options, middleware);
}
```